shell/platform/darwin/ios/framework/Source/FlutterViewController.mm

```diff
@@ -1069,56 +1097,60 @@ - (void)keyboardWillBeHidden:(NSNotification*)notification {
   [self updateViewportMetrics];
 }
 
-- (void)dispatchPresses:(NSSet<UIPress*>*)presses API_AVAILABLE(ios(13.4)) {
+- (void)handlePressEvent:(NSSet<UIPress*>*)presses
+               withEvent:(UIPressesEvent*)event
+              nextAction:(void (^)())next API_AVAILABLE(ios(9.0)) {
   if (@available(iOS 13.4, *)) {
-    for (UIPress* press in presses) {
-      if (press.key == nil || press.phase == UIPressPhaseStationary ||
-          press.phase == UIPressPhaseChanged) {
-        continue;
-      }
-      NSMutableDictionary* keyMessage = [[@{
-        @"keymap" : @"ios",
-        @"type" : @"unknown",
-        @"keyCode" : @(press.key.keyCode),
-        @"modifiers" : @(press.key.modifierFlags),
-        @"characters" : press.key.characters,
-        @"charactersIgnoringModifiers" : press.key.charactersIgnoringModifiers
-      } mutableCopy] autorelease];
-
-      if (press.phase == UIPressPhaseBegan) {
-        keyMessage[@"type"] = @"keydown";
-      } else if (press.phase == UIPressPhaseEnded || press.phase == UIPressPhaseCancelled) {
-        keyMessage[@"type"] = @"keyup";
-      }
-
-      [[_engine.get() keyEventChannel] sendMessage:keyMessage];
+    bool __block wasHandled = false;
+    KeyEventCompleteCallback completeCallback =
+        ^void(bool handled, UIPress* press, UIPressesEvent* event) {
+          wasHandled = handled;
+          CFRunLoopStop(CFRunLoopGetCurrent());
+        };
+    [_keyboardManager handlePresses:presses withEvent:event complete:completeCallback];
+    CFRunLoopRun();
```

Comment:
Tangentially related: If it turns out that nested runloop causes problems here (i.e. we're holding it for too long and iOS delivers another event when you don't expect it), the solution could be similar to what I used for synchronized resizing in the linux embedder: [fl_task_runner.h](https://github.com/flutter/engine/pull/25884/files#diff-8bffa3ea93e572d33adc0a6e19d2145dbee901fb892f124341e9987ab81a21d1) [fl_task_runner.cc](https://github.com/flutter/engine/pull/25884/files#diff-8c19fc6644aae30fab91a913c6042841dcb2ff3adf35d71a0703cc9cc409422d). The FlTaskRunner class can hold the main thread while processing Flutter tasks at the same time. Adapting it to `CFRunLoop` should be fairly straightforwrd.